[BUG] Query failed with java.lang.NullPointerException when doing GpuSubqueryBroadcastExec

One customer hit this issue that one query ailed with java.lang.NullPointerException when doing GpuSubqueryBroadcastExec.

**Env:**
Dataproc 2.0
Spark 3.1.3
rapids-4-spark_2.12-22.08.0-20220803.095308-37.jar

**Symptom**
There is no job/stage/task error in Spark eventlog.
The full stacktrace is from Spark driver log:
```
Caused by: java.lang.NullPointerException
	at ai.rapids.cudf.HostColumnVectorCore.getStartListOffset(HostColumnVectorCore.java:275)
	at ai.rapids.cudf.HostColumnVectorCore.getUTF8(HostColumnVectorCore.java:365)
	at com.nvidia.spark.rapids.RapidsHostColumnVectorCore.getUTF8String(RapidsHostColumnVectorCore.java:183)
	at org.apache.spark.sql.vectorized.ColumnarBatchRow.getUTF8String(ColumnarBatch.java:220)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$5(GpuSubqueryBroadcastExec.scala:199)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)
	at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)
	at scala.collection.mutable.ArrayBuilder$ofRef.$plus$plus$eq(ArrayBuilder.scala:109)
	at scala.collection.mutable.ArrayBuilder$ofRef.$plus$plus$eq(ArrayBuilder.scala:66)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.mutable.ArrayOps$ofRef.flatMap(ArrayOps.scala:198)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$3(GpuSubqueryBroadcastExec.scala:197)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:55)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:53)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.withResource(GpuSubqueryBroadcastExec.scala:142)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$2(GpuSubqueryBroadcastExec.scala:196)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withExecutionId$1(SQLExecution.scala:139)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.withExecutionId(SQLExecution.scala:137)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$1(GpuSubqueryBroadcastExec.scala:182)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

I can repro locally with a minimum repro.
Env:
Spark 3.2.1 standalone cluster
Spark RAPIDS 22.06 GA jar
Apache Hive 3.1.2  `spark.sql.hive.metastore.version 3.1`


Reproduce steps:
```
create table debugdb.dim (
`id` string)
PARTITIONED BY (                                   
  `dt` date,                            
  `hr` string,                          
  `mins` string)                         
Stored as parquet;
insert into debugdb.dim(id,dt,hr,mins) values ('somevalue', date('2022-01-01'), '11', '59');


CREATE TABLE debugdb.fact(
  `id` string)                     
PARTITIONED BY (                                   
  `dt` date,                                 
  `hr` string,                               
  `mins` string)                              
Stored as parquet;

insert into debugdb.fact(id,dt,hr,mins)
select 'somevalue', to_date('2022-01-01'), '11', '59' from tpcds.customer;

select count(*) as cnt
from debugdb.fact csd
left join (select *,
from_utc_timestamp(date_format(concat(string(dt),' ',hr,':',mins,':','00'),'yyyy-MM-dd HH:mm:ss.SSS'),'UTC') as ts
from debugdb.dim 
) tmp
on  csd.hr = tmp.hr
and csd.mins = tmp.mins
and csd.dt = tmp.dt
where unix_timestamp(tmp.ts) > unix_timestamp(CURRENT_TIMESTAMP) - 99999*3600 ;
```

Reproduced stacktrace:
```
Caused by: java.lang.NullPointerException
	at ai.rapids.cudf.HostColumnVectorCore.getStartListOffset(HostColumnVectorCore.java:275)
	at ai.rapids.cudf.HostColumnVectorCore.getUTF8(HostColumnVectorCore.java:365)
	at com.nvidia.spark.rapids.RapidsHostColumnVectorCore.getUTF8String(RapidsHostColumnVectorCore.java:183)
	at org.apache.spark.sql.vectorized.ColumnarBatchRow.getUTF8String(ColumnarBatch.java:220)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$5(GpuSubqueryBroadcastExec.scala:199)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)
	at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)
	at scala.collection.mutable.ArrayBuilder$ofRef.$plus$plus$eq(ArrayBuilder.scala:109)
	at scala.collection.mutable.ArrayBuilder$ofRef.$plus$plus$eq(ArrayBuilder.scala:66)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.mutable.ArrayOps$ofRef.flatMap(ArrayOps.scala:198)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$3(GpuSubqueryBroadcastExec.scala:197)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:55)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:53)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.withResource(GpuSubqueryBroadcastExec.scala:142)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$2(GpuSubqueryBroadcastExec.scala:196)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withExecutionId$1(SQLExecution.scala:139)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.withExecutionId(SQLExecution.scala:137)
	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$1(GpuSubqueryBroadcastExec.scala:182)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

Note: This is a typical DPP feature. 
If I disable DPP, the GPU run works fine: ` set spark.sql.optimizer.dynamicPartitionPruning.enabled=false;`

--- a/integration_tests/src/main/python/dpp_test.py
+++ b/integration_tests/src/main/python/dpp_test.py
@@ -14,7 +14,7 @@
 
 import pytest
 
-from asserts import assert_cpu_and_gpu_are_equal_collect_with_capture
+from asserts import assert_cpu_and_gpu_are_equal_collect_with_capture, assert_gpu_and_cpu_are_equal_collect
 from conftest import spark_tmp_table_factory, is_databricks_runtime
 from data_gen import *
 from marks import ignore_order, allow_non_gpu
@@ -269,3 +269,31 @@ def test_dpp_like_any(spark_tmp_table_factory, store_format, aqe_enabled):
         lambda spark: spark.sql(statement),
         exist_classes='DynamicPruningExpression,GpuSubqueryBroadcastExec,ReusedExchangeExec',
         conf=dict(_exchange_reuse_conf + [('spark.sql.adaptive.enabled', aqe_enabled)]))
+
+# Test handling DPP expressions from a HashedRelation that rearranges columns
+@pytest.mark.parametrize('aqe_enabled', [
+    'false',
+    pytest.param('true', marks=pytest.mark.skipif(is_before_spark_320(),
+                                                  reason='Only in Spark 3.2.0+ AQE and DPP can be both enabled'))
+], ids=idfn)
+@pytest.mark.skipif(is_databricks_runtime(), reason='DPP is not supported on Databricks runtime')
+def test_dpp_from_swizzled_hash_keys(spark_tmp_table_factory, aqe_enabled):
+    dim_table = spark_tmp_table_factory.get()
+    fact_table = spark_tmp_table_factory.get()
+    def setup_tables(spark):
+        spark.sql("CREATE TABLE {}(id string) PARTITIONED BY (dt date, hr string, mins string) STORED AS PARQUET".format(dim_table))
+        spark.sql("INSERT INTO {}(id,dt,hr,mins) values ('somevalue', date('2022-01-01'), '11', '59')".format(dim_table))
+        spark.sql("CREATE TABLE {}(id string)".format(fact_table) +
+                  " PARTITIONED BY (dt date, hr string, mins string) STORED AS PARQUET")
+        spark.sql("INSERT INTO {}(id,dt,hr,mins)".format(fact_table) +
+                  " SELECT 'somevalue', to_date('2022-01-01'), '11', '59'")
+    with_cpu_session(setup_tables)
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark : spark.sql("SELECT COUNT(*) AS cnt FROM {} f".format(fact_table) +
+                                 " LEFT JOIN (SELECT *, " +
+                                 " date_format(concat(string(dt),' ',hr,':',mins,':','00'),'yyyy-MM-dd HH:mm:ss.SSS') AS ts" +
+                                 " from {}) tmp".format(dim_table) +
+                                 " ON f.hr = tmp.hr AND f.dt = tmp.dt WHERE tmp.ts < CURRENT_TIMESTAMP"),
+        conf=dict(_dpp_conf + [('spark.sql.adaptive.enabled', aqe_enabled),
+                               ("spark.rapids.sql.castStringToTimestamp.enabled", "true"),
+                               ("spark.rapids.sql.hasExtendedYearValues", "false")]))

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
@@ -28,9 +28,11 @@ import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference, BoundReference, Cast, Expression, NamedExpression, UnsafeProjection}
 import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.catalyst.plans.physical.IdentityBroadcastMode
 import org.apache.spark.sql.execution.{BaseSubqueryExec, SparkPlan, SQLExecution, SubqueryBroadcastExec}
 import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec
 import org.apache.spark.sql.execution.exchange.BroadcastExchangeExec
+import org.apache.spark.sql.execution.joins.HashedRelationBroadcastMode
 import org.apache.spark.util.ThreadUtils
 
 
@@ -134,7 +136,29 @@ class GpuSubqueryBroadcastMeta(
   override def convertToCpu(): SparkPlan = s
 
   override def convertToGpu(): GpuExec = {
-    GpuSubqueryBroadcastExec(s.name, s.index, s.buildKeys, broadcastBuilder())
+    GpuSubqueryBroadcastExec(s.name, s.index, s.buildKeys, broadcastBuilder())(
+      getBroadcastModeKeyExprs)
+  }
+
+  /** Extract the broadcast mode key expressions if there are any. */
+  private def getBroadcastModeKeyExprs: Option[Seq[Expression]] = {
+    val broadcastMode = s.child match {
+      case b: BroadcastExchangeExec =>
+        b.mode
+      case a: AdaptiveSparkPlanExec =>
+        SparkShimImpl.getAdaptiveInputPlan(a) match {
+          case b: BroadcastExchangeExec =>
+            b.mode
+          case _ =>
+            throw new AssertionError("should not reach here")
+        }
+    }
+
+    broadcastMode match {
+      case HashedRelationBroadcastMode(keys, _) => Some(keys)
+      case IdentityBroadcastMode => None
+      case m => throw new UnsupportedOperationException(s"Unknown broadcast mode $m")
+    }
   }
 }
 
@@ -143,7 +167,10 @@ case class GpuSubqueryBroadcastExec(
     name: String,
     index: Int,
     buildKeys: Seq[Expression],
-    child: SparkPlan) extends BaseSubqueryExec with GpuExec with ShimUnaryExecNode {
+    child: SparkPlan)(modeKeys: Option[Seq[Expression]])
+    extends BaseSubqueryExec with GpuExec with ShimUnaryExecNode {
+
+  override def otherCopyArgs: Seq[AnyRef] = modeKeys :: Nil
 
   // As `SubqueryBroadcastExec`, `GpuSubqueryBroadcastExec` is only used with `InSubqueryExec`.
   // No one would reference this output, so the exprId doesn't matter here. But it's important to
@@ -168,7 +195,7 @@ case class GpuSubqueryBroadcastExec(
 
   override def doCanonicalize(): SparkPlan = {
     val keys = buildKeys.map(k => QueryPlan.normalizeExpressions(k, child.output))
-    GpuSubqueryBroadcastExec("dpp", index, keys, child.canonicalized)
+    GpuSubqueryBroadcastExec("dpp", index, keys, child.canonicalized)(modeKeys)
   }
 
   @transient
@@ -185,18 +212,28 @@ case class GpuSubqueryBroadcastExec(
         val serBatch = child.executeBroadcast[SerializeConcatHostBuffersDeserializeBatch]()
 
         // Creates projection to extract target field from Row, as what Spark does.
-        val rowProject = {
-          val extractKey = BoundReference(
-            index, buildKeys(index).dataType, buildKeys(index).nullable)
-          UnsafeProjection.create(extractKey)
+        // Note that unlike Spark, the GPU broadcast data has not applied the key expressions from
+        // the HashedRelation, so that is applied here if necessary to ensure the proper values
+        // are being extracted. The CPU already has the key projections applied in the broadcast
+        // data and thus does not have similar logic here.
+        val broadcastModeProject = modeKeys.map { keyExprs =>
+          val keyExpr = keyExprs(index)
+          UnsafeProjection.create(keyExpr)
         }
+
+        // Use the single output of the broadcast mode projection if it exists
+        val rowProjectIndex = if (broadcastModeProject.isDefined) 0 else index
+        val rowProject = UnsafeProjection.create(
+          BoundReference(rowProjectIndex, buildKeys(index).dataType, buildKeys(index).nullable))
 
         // Deserializes the batch on the host. Then, transforms it to rows and performs row-wise
         // projection. We should NOT run any device operation on the driver node.
         val result = withResource(serBatch.value.hostBatches) { hostBatches =>
           hostBatches.flatMap { cb =>
-            cb.rowIterator().asScala
-              .map(rowProject(_).copy().asInstanceOf[InternalRow])
+            cb.rowIterator().asScala.map { row =>
+              val broadcastRow = broadcastModeProject.map(_(row)).getOrElse(row)
+              rowProject(broadcastRow).copy().asInstanceOf[InternalRow]
+            }
           }
         }
 

